packages/grid/x-grid/src/tests/state.DataGridPro.test.tsx

```diff
@@ -62,6 +62,7 @@ describe('<DataGridPro /> - State', () => {
           ...prev,
           sorting: { ...prev.sorting, sortModel: [{ field: 'brand', sort: 'asc' }] },
         }));
+        apiRef.current.applySorting();
```

Comment:
Why do we need to add this imperative action? I was under the assumption that the sorting should be applied when the state changes, e.g. with a controller prop or imperative call.

Reply:
As I understand it, the sorting and filter are not applied systematically when their state change.
We have a manual `applySorting` / `applyFilter` in the update functions.

And I think before, it was doing something like this

1. Render 1 of the Grid
2. useEffect of the Grid in useGridRows, the rows are stored in state + useEffect of the Grid in useGridSorting listening to props.rows, the sorting is applied
3. useEffect of the test runs
4. useEffect of the Grid in useGridSorting listening to the rowCount selector due to the state rows change, the sorting is reapplied

With my last commit, I removed the 2nd sorting application that was after you update the state in the test

One way to confirm is to delay the state update of the test, and it fails on master

![image](https://user-images.githubusercontent.com/3309670/132053855-8c6b87e9-89c2-4d2f-ac34-ab99cb9f86a5.png)


That's typically why I want to reduce as much as possible these async state update cascades, they make things work for the wrong reason in some tests.

Reply:
Do we want the user to update the state manually ? Is it something we see as normal or is it something they should do at their own risk ?

Because most hooks have a method to cleanly update, method that can rerun other behaviors (`applySorting` for instance) and `forceRender`.

By default, a `apiRef.current.setState` does not
- rerender
- apply side effects like `applySorting`
- apply the control state constraints

For me we should at the very least migrate the `setGridState` code of `useGridState` inside `apiRef.current.setState`
`apiRef.current.setState` feels like legacy to me in its current shape

Reply:
Good points, I don't think that my initial question challenges that we can merge it.

I think that the main question for me is should the application of the sorting directly depends on the state to update, or should it depends on higher level signals, like the fact that a `sortModel` prop changes.

Reply:
>  I think what this test is doing is a "DataGrid anti-pattern".

I agree on this
I'll do another PR

Reply:
For the merge of `rowsSet` and `rowsUpdate`, I agree, that can be done later one :+1: